### PR TITLE
Some changes to calls to Docker API for podman compatibility

### PR DESCRIPTION
**Pull request by MatMaul**
- Podman expects the network name specified when launching the container to match the name used when creating it even when the network ID is specified, the API spec is unclear about that
- Adding labels when committing a container needs to use `changes` query parameter instead of a POST json, which is unspecified
- Add a config to specify the hostname where Complement on the host can be reached when inside a container. Podman uses `host.containers.internal`

### CODE REVIEW

File: internal/docker/builder.go [hunk 1]
```diff
@@ -328,6 +328,10 @@ func (d *Builder) construct(bprint b.Blueprint) (errs []error) {
 		// Log again so we can see the timings.
 		d.log("%s: Stopped container: %s", res.contextStr, res.containerID)
 
+		for _, c := range changes {
+			d.log(c)
+		}
```
Comment (reivilibre) [inline internal/docker/builder.go:333]: debug code left in? If you want to keep this, probably want to prefix the log line with something meaningful

**Review by reivilibre — APPROVED** [review]
Looks reasonable to me

File: internal/docker/builder.go (hunk 1 above)
Comment (MatMaul) [reply internal/docker/builder.go]: thanks, fixed :)

**Comment by MatMaul** [comment]
It seems like the hack [here](https://github.com/matrix-org/complement/blob/main/internal/docker/deployer.go#L187) was only needed because of the fact that the network name was not the one used when launching the container.

File: internal/docker/builder.go [hunk 2]
```diff
@@ -292,32 +289,32 @@ func (d *Builder) construct(bprint b.Blueprint) (errs []error) {
...
 			for k, v := range accessTokens {
-				labels["access_token_"+k] = v
+				changes = appendLabelToChanges(changes, "access_token_"+k, v)
 			}
 		}
 
 		deviceIDs := runner.DeviceIDs(res.homeserver.Name)
 		for userID, deviceID := range deviceIDs {
-			labels["device_id"+userID] = deviceID
+			changes = appendLabelToChanges(changes, "device_id"+userID, deviceID)
```
Comment (kegsay) [inline internal/docker/builder.go:311]: This presumably is going to result in duplicate keys if there are two users with the same ID? Whereas previously we would just take the last value.

File: internal/docker/builder.go [hunk 3]
```diff
@@ -336,9 +333,7 @@ func (d *Builder) construct(bprint b.Blueprint) (errs []error) {
 			Author:    "Complement",
 			Pause:     true,
 			Reference: "localhost/complement:" + res.contextStr,
-			Config: &container.Config{
-				Labels: labels,
-			},
+			Changes:   changes,
```
Comment (kegsay) [inline internal/docker/builder.go:336]: I'd prefer it if we kept the map of `labels` and then just called `toChanges(labels) []string` or something.

File: internal/docker/deployer.go [hunk 4]
```diff
@@ -184,14 +184,6 @@ func (d *Deployer) Restart(hsDep *HomeserverDeployment, cfg *config.Complement)
 		return fmt.Errorf("Restart: Failed to stop container %s: %s", hsDep.ContainerID, err)
 	}
 
-	// Remove the container from the network. If we don't do this,
-	// (re)starting the container fails with an error like
-	// "Error response from daemon: endpoint with name complement_fed_1_fed.alice.hs1_1 already exists in network complement_fed_alice".
-	err = d.Docker.NetworkDisconnect(ctx, d.networkID, hsDep.ContainerID, false)
-	if err != nil {
-		return fmt.Errorf("Restart: Failed to disconnect container %s: %s", hsDep.ContainerID, err)
-	}
```
Comment (kegsay) [inline internal/docker/deployer.go:193]: Why was this removed?

File: internal/docker/deployer.go [hunk 5]
```diff
@@ -283,7 +275,7 @@ func deployImage(
 		Mounts:     mounts,
 	}, &network.NetworkingConfig{
 		EndpointsConfig: map[string]*network.EndpointSettings{
-			contextStr: {
+			"complement_" + pkgNamespace + "_" + blueprintName: {
```
Comment (kegsay) [inline internal/docker/deployer.go:278]: Why aren't we using the `contextStr` anymore?

File: internal/docker/labels.go [hunk 6]
```diff
@@ -33,7 +33,7 @@ func asIDToRegistrationFromLabels(labels map[string]string) map[string]string {
 	asMap := make(map[string]string)
 	for k, v := range labels {
 		if strings.HasPrefix(k, "application_service_") {
-			asMap[strings.TrimPrefix(k, "application_service_")] = v
+			asMap[strings.TrimPrefix(k, "application_service_")] = strings.Replace(v, "\\n", "\n", -1)
```
Comment (kegsay) [inline internal/docker/labels.go:36]: Why did you change the format of `\n` and then do a replace over it?

**Review by kegsay — CHANGES_REQUESTED** [review]
In theory LGTM, but I have outstanding questions on why seemingly unrelated things were modified. The most dangerous is the blunt removal of network disconnection code, which the comments explained why it existed, and yet this PR just removes without any reason...

File: internal/docker/labels.go (hunk 6 above)
Comment (MatMaul) [reply internal/docker/labels.go:36]: Because the `change` is a Dockerfile directive and it seems like it can't be on several lines.
This allows generating this kind of change `LABEL "mylabel"="multiline\nvalue"`.

File: internal/docker/builder.go (hunk 2 above)
Comment (MatMaul) [reply internal/docker/builder.go:311]: Exact, I'll do what you suggested (use `labels` and then convert to `changes`) and this problem should be solved.

File: internal/docker/deployer.go (hunk 5 above)
Comment (MatMaul) [reply internal/docker/deployer.go:278]: This needs to be the name of the network, which is not the case here.

We could also use the contextStr as the name of the network in `createNetworkIfNotExists`, or makes `createNetworkIfNotExists` returns both the ID and the network name and store both in `Deployer`.

The latter looks like the cleaner way to do it so leaning towards that.

File: internal/docker/deployer.go (hunk 4 above)
Comment (MatMaul) [reply internal/docker/deployer.go:193]: Following your edit in the top comment, do you have more question on this one ?

File: internal/docker/deployer.go (hunk 5 above)
Comment (MatMaul) [reply internal/docker/deployer.go]: Actually it seems like we can just use the name, ID is optional, at least with Podman.
I've push a change to return and use the network name instead of the ID, let's see if Docker is happy.

File: internal/docker/labels.go (hunk 6 above)
Comment (MatMaul) [reply internal/docker/labels.go]: I added comments in the code regarding that for clarity.

File: internal/docker/deployer.go (hunk 4 above)
Comment (kegsay) [reply internal/docker/deployer.go:193]: Nope!

File: internal/config/config.go [hunk 7]
```diff
@@ -78,6 +78,9 @@ type Complement struct {
 	CAPrivateKey  *rsa.PrivateKey
 
 	BestEffort bool
+
+	// The hostname of Complement from the perspective of a Homeserver running inside a container
+	HostnameRunningComplement string
```
Comment (kegsay) [inline internal/config/config.go:83]: Please update the docstring to match other env vars in this struct. Use this format so we can automatically generate docs:
```go
// Name: COMPLEMENT_HOSTNAME_RUNNING_COMPLEMENT
// Default: host.docker.internal
// Description: Some useful description here,
// multi-line is fine.
```

Please be thorough and explain:
 - Why you might want to set this.
 - What value you might want to set this to in Podman as an example.

File: internal/docker/builder.go [hunk 8]
```diff
@@ -351,11 +346,19 @@ func (d *Builder) construct(bprint b.Blueprint) (errs []error) {
 	return errs
 }
 
+func toChanges(labels map[string]string) []string {
+	changes := make([]string, 0)
```
Comment (kegsay) [inline internal/docker/builder.go:350]: `var changes []string` is shorter, or `changes := make([]string, 0, len(labels))` to preallocate a capacity of the underlying array.

File: internal/docker/builder.go [hunk 9]
```diff
@@ -351,11 +346,19 @@ func (d *Builder) construct(bprint b.Blueprint) (errs []error) {
...
+	for k, v := range labels {
+		changes = append(changes, fmt.Sprintf("LABEL \"%s\"=\"%s\"", k, v))
```
Comment (kegsay) [inline internal/docker/builder.go:352]: Please document assumptions you're making here. Not all characters will be supported in this format, no? Which ones would be problematic?

**Review by kegsay — APPROVED** [review]
Otherwise LGTM

**Comment by MatMaul** [comment]
@kegsay thanks for the review, I'll let you do a final round and merge. My Fedora is happy :) .

**Review by kegsay — APPROVED** [review]
